Adding SPRegion region class

dkeeney opened the pull request:
This is a PR for what I have for the SPRegion plugin.   It is mostly complete.

dkeeney commented:
```
c:\projects\nupic-cpp\build\thirdparty\eigen\eigen-src\eigen\src\core\products\generalmatrixmatrixtriangular.h(300): error C2244: 'Eigen::TriangularViewImpl<_MatrixType,_Mode,Eigen::Dense>::_assignProduct': unable to match function definition to an existing
```
Don't know where this came from.  Something got into the master I think.

breznak commented on src/test/unit/regions/RegionTestUtilities.hpp:
```diff
@@ -0,0 +1,55 @@
+ 
+#include <iostream>
+#include <nupic/engine/Region.hpp>
+#include <nupic/algorithms/BacktrackingTMCpp.hpp>
```
Does that have to be here?

dkeeney replied:
> #include <nupic/algorithms/BacktrackingTMCpp.hpp>

This module is shared with the BacktrackingTMRegion class and TMRegion class unit tests.  Both are coming.

breznak commented on src/test/unit/regions/RegionTestUtilities.hpp:
```diff
@@ -0,0 +1,55 @@
+using namespace nupic;
+namespace testing 
+{
+typedef std::shared_ptr<Region> Region_Ptr_t;
```
Nit, consider removing the typedef and use the type directly

breznak commented on src/test/unit/regions/SPRegionTest.cpp:
```diff
@@ -0,0 +1,396 @@
+#include <nupic/os/Timer.hpp>
+#include <nupic/os/Directory.hpp>
+#include <nupic/engine/YAMLUtils.hpp>
+#include <nupic/regions/SPRegion.hpp>
```
Quite a lot of includes, can some already be covered by SPRegion hpp

breznak commented on src/test/unit/regions/RegionTestUtilities.cpp:
```diff
@@ -0,0 +1,529 @@
+
+#define VERBOSE                                                                \
+  if (verbose)                                                                 \
+  std::cerr << "[          ] "
```
Can this cause error that the define is not on a single line?

dkeeney replied:
> Can this cause error that the define is not on a single line?

The '\' on the far right makes it one line.
But I can make it a single line.  I think I did other places.

breznak commented:
I’m reviewing on phone, so sorry for shorter replies. I’ll be back on Wed, but I’m trying to read mail

dkeeney commented on src/nupic/regions/SPRegion.cpp:
```diff
@@ -0,0 +1,1266 @@
+
+    // Direct logging of SP non-zero inputs if requested
+    if (!logPathInput_.empty()) {
+      FILE *fp = fopen(logPathInput_.c_str(), "a");
```
This is basically a direct port from the .py code.  So what you see are things that the .py supports.

The file open could be replaced with a stream I guess.

breznak replied:
> The file open could be replaced with a stream I guess. 

I would even go and remove this logging possibility (?)

dkeeney replied:
For TM's I commented out the logging and the parameters for it. (keeping the code just in case it is important to someone). You could do the same for SP.

breznak replied:
commended out

dkeeney commented on src/nupic/regions/SPRegion.hpp:
```diff
@@ -0,0 +1,146 @@
+
+	
+	private:
+		SPRegion();  // empty constructor not allowed
```
yes, = delete would be better.

dkeeney commented on src/nupic/regions/SPRegion.hpp:
```diff
@@ -0,0 +1,146 @@
+      UInt spVerbosity;
+      bool wrapAround;
+      bool learningMode;
+      bool inferenceMode;  // This mode not implemented.
```
> Inference, as process data without learning, is not implemented?

Was not in the .py that I ported from.

breznak replied:
What I mean is, for SP, inference is producing SDR from the inputs without learning the columns. 
Learning is the same, but with learning on.

dkeeney commented on src/nupic/regions/SPRegion.hpp:
```diff
@@ -0,0 +1,146 @@
+    std::string logPathOutput_;  
+    std::string logPathOutputDense_; // can be set and get but not used for anything.
+
+    SpatialPooler* sp_;
```
> Smart pointer, 

Actually no.  This is allocated here but never shared.  Perhaps one of the other flavors of smart pointers.

dkeeney replied:
> Smart pointer,

Sorry.  Yes, should use smart pointer.  No not a shared_ptr.

breznak replied:
or even no pointer, but a direct instance/member?

breznak replied:
added this, produces some errors with mis-match types

breznak commented:
On Fri, 1 Feb 2019 at 21:10, David Keeney <notifications@github.com> wrote:

> *@dkeeney* commented on this pull request.
> ------------------------------
>
> In src/nupic/regions/SPRegion.cpp
> <https://github.com/htm-community/nupic.cpp/pull/244#discussion_r253184470>
> :
>
> > +    // Direct logging of SP non-zero outputs if requested
> +    if (!logPathOutput_.empty()) {
> +      FILE *fp = fopen(logPathOutput_.c_str(), "a");
> +      if (fp) {
> +        for (int i = 0; i < size; i++) {
> +          if (outputVector[i])
> +            fprintf(fp, "%d ", i);
> +        }
> +        fprintf(fp, "\n");
> +        fclose(fp);
> +      }
> +    }
> +
> +    // Direct logging of SP non-zero inputs if requested
> +    if (!logPathInput_.empty()) {
> +      FILE *fp = fopen(logPathInput_.c_str(), "a");
>
> This is basically a direct port from the .py code. So what you see are
> things that the .py supports.
>
> The file open could be replaced with a stream I guess.
>
i meant removing such functionality?

—
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/htm-community/nupic.cpp/pull/244#discussion_r253184470>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACENX7m4BKHKblqaZ6IpyzF00ZrHmbmUks5vJJ8lgaJpZM4aer1z>
> .
>
-- 
Marek Otahal :o)

dkeeney commented:
Let me give you a quick overview of a plugin (a region implementation).
All plugins must subclass RegionImpl. 

implement the following:
- constructor, myRegion(const ValueMap &params, Region *region)   params will contain all passed in arguments. Extract what you need.  The ValueMap goes away when you return.
- getters and setters using specified format. parameters other than constructor args.  Only override those you need.  
- createSpec()   required.  This is a STATIC function so cannot do overrides but required. Returns a static structure that describes the plugin.  See API docs.
- serialize( )      - serialize yourself and the algorithm you are driving
- deserialize( )  - load your algorithm and yourself
- initialize( )   - instantiate your algorithm with args from and initialize it.
- compute( )  - Perform one iteration of execution of the algorithm
- executeCommand( ) - For rare special cases perform some action.
- getNodeOutputElementCount(output_name)  - return the width of the output buffer for this output.

That is it.
Have fun.

dkeeney commented on external/CMakeLists.txt:
```diff
@@ -57,7 +57,7 @@ set(EP_BASE ${CMAKE_BINARY_DIR})
   set(EXTERNAL_CXX_FLAGS /EHsc "$<$<CONFIG:Release>:/MD>" "$<$<CONFIG:Debug>:/MDd>")
 else()
-  set(EXTERNAL_CXX_FLAGS -fPIC)
+	set(EXTERNAL_CXX_FLAGS -fPIC)
```
tabs as spaces?

dkeeney commented on src/nupic/regions/SPRegion.cpp:
```diff
@@ -121,13 +120,13 @@ void SPRegion::initialize() {
     args_.potentialRadius = args_.inputWidth;
 
-  sp_ = new SpatialPooler(
+  sp_ = std::unique_ptr<SpatialPooler>(new SpatialPooler(
```
👍 
Keep in mind that this makes SPRegion non-copyable.  I think that is ok.

breznak replied:
The smart pointer leads to some more errors, 

> [          ] Setup Network; add 3 regions and 2 links.
ERR:  CHECK FAILED: "srcOutput->getDataType() == destInput->getDataType()" Network::link -- Mismatched data types.Real32 != UInt32 [/home/travis/build/htm-community/nupic.cpp/src/nupic/engine/Network.cpp line 312]
unknown file: Failure
C++ exception with description "CHECK FAILED: "srcOutput->getDataType() == destInput->getDataType()" Network::link -- Mismatched data types.Real32 != UInt32" thrown in the test body.
[  FAILED  ] SPRegionTest.testLinking (0 ms 

W/o it the tests cleanly pass.

dkeeney replied:
Oh, the network in master does not have my changes so types must still match.

breznak replied:
> network in master does not have my changes so types must still match.

Can you point me where the datatype of the field (srcOutput, destInput) is defined?

dkeeney commented on src/nupic/regions/SPRegion.cpp:
```diff
@@ -172,6 +171,8 @@ void SPRegion::compute() {
+
     size_t size = sp_->getNumColumns();
     // Direct logging of SP non-zero outputs if requested
     if (!logPathOutput_.empty()) {
```
I think we can remove logging.

breznak replied:
ok, I'll remove even the comments.

breznak commented:
After merging the SDR-NetworkAPI PR #234 , what is the replacement for `Array.nonZero()` or what construction should be used? @dkeeney 

```
 -    // Create a NonZero array from the indexes of non-zero values of the local array.
-    Array nonZero() const {
-      Array a(NTA_BasicType_UInt32);
-      ArrayBase::NonZero(a);
-      return a; 
```

From compilation
```
 home/mmm/devel/HTM/htm-community/nupic.cpp/src/nupic/regions/SPRegion.cpp: In member function ‘virtual void nupic::SPRegion::getParameterArray(const string&, nupic::Int64, nupic::Array&)’:
/home/mmm/devel/HTM/htm-community/nupic.cpp/src/nupic/regions/SPRegion.cpp:929:27: error: ‘const class nupic::Array’ has no member named ‘nonZero’
nzInput_ = incoming.nonZero();
^~~~~~~
/home/mmm/devel/HTM/htm-community/nupic.cpp/src/nupic/regions/SPRegion.cpp:936:26: error: ‘const class nupic::Array’ has no member named ‘nonZero’
nzOutput_ = output.nonZero();
^~~~~~~
```

breznak commented on src/nupic/regions/SPRegion.cpp:
```diff
@@ -0,0 +1,1239 @@
+    if (!nzInputValid_) {
+      const Array &incoming = getInput("bottomUpIn")->getData(); //transform incoming to sparse format
+      const SDR nonZero = *incoming.getSDR();
+      nzInput_ = Array(nonZero);
```
is this the correct way to replace Array::nonZero() ?

dkeeney replied:
What this is trying to do is if spOutputNonZeros parameter is requested, return the output in sparse format.  So... is the output buffer using SDR's for the SP algorithm yet?  If it is then just get the SDR that is returned from the SP algorithm, and call getFlatSparse( ) which will return a vector that you can return.  In order to get the plugin to expect Arrays with an SDR and to output Arrays containing SDR's you need to tell NetworkAPI that you are expecting them by setting NTA_BasicType_SDR as the type in the Spec for the bottomUpOut output and bottomUpIn input.  The link will then create them as SDR type Arrays.

If you are not using SDR's in the algorithm yet, then you will need to create an SDR, populate it with the bottomUpOut buffer contents, then you can call the getFlatSparse( ) to get the FlatSparse buffer to return..

Actually, if we used an SDR in the Array for the bottomUpOut output, the spOutputNonZeros parameter is obsolete because the caller can get it from the output.

dkeeney replied:
Hmmm, the spOutputNonZeros parameter is expecting to return an Array, not a vector.  If the bottomUpOut is  an NTA_BasicType_SDR type, then all you need to do is return the bottomUpOut Array.

dkeeney replied:
If you don't have an SDR type for bottomUpOut (i.e. the algorithm is not setup to handle them yet) then you can:
```
const Array& bottomUpOut = getOutput("bottomUpOut")->getData();
return bottomUpOut.get_as(NTA_BasicType_SDR);
```

breznak replied:
I'm lost here, when you have time, can you have a look, please?

dkeeney replied:
Try this:
```
// copy the contents of the requested array into the caller's array.
// Allocate the buffer if one is not provided.  Convert data types if needed.
// The returned Array will be type NTA_BasicType_SDR which contains both dense 
// and sparse formats.
void SPRegion::getParameterArray(const std::string &name, Int64 index, Array &array) {
  if (name == "spatialPoolerInput") {
      return getInput("bottomUpIn")->getData().get_as(NTA_BasicType_SDR);
  } else if (name == "spatialPoolerOutput") {
     return getOutput("bottomUpOut")->getData().get_as(NTA_BasicType_SDR);
  } else if (name == "spInputNonZeros") {
      return getInput("bottomUpIn")->getData().get_as(NTA_BasicType_SDR);
  } else if (name == "spOutputNonZeros") {
     return getOutput("bottomUpOut")->getData().get_as(NTA_BasicType_SDR);
  }
  else {
    this->RegionImpl::getParameterArray(name, index, array);
  }
}
```

dkeeney replied:
oops, its not reutnred. Its an argument.
```
// copy the contents of the requested array into the caller's array.
// Allocate the buffer if one is not provided.  Convert data types if needed.
// The returned Array will be type NTA_BasicType_SDR which contains both dense 
// and sparse formats.
void SPRegion::getParameterArray(const std::string &name, Int64 index, Array &array) {
  if (name == "spatialPoolerInput") {
      array = getInput("bottomUpIn")->getData().get_as(NTA_BasicType_SDR);
  } else if (name == "spatialPoolerOutput") {
     array = getOutput("bottomUpOut")->getData().get_as(NTA_BasicType_SDR);
  } else if (name == "spInputNonZeros") {
      array = getInput("bottomUpIn")->getData().get_as(NTA_BasicType_SDR);
  } else if (name == "spOutputNonZeros") {
     array = getOutput("bottomUpOut")->getData().get_as(NTA_BasicType_SDR);
  }
  else {
    this->RegionImpl::getParameterArray(name, index, array);
  }
}
```

dkeeney replied:
In the unit tests; SPRegionTest, line 330 to 336 is expecting NTA_BasicType_UInt32 so it will fail.  Changing it to NTA_BasicType_SDR will fix that but 
compareParameterArrays( ) is only going to compare the dense format.

So, since we ripped out the NTA_BasicType_Sparse, we don't have a way to represent the sparse format as anything other than an SDR.   I guess we could change the test to also compare the sparse content but that capability is already covered in the SDR tests.  So perhaps comparing the dense format is sufficient.

dkeeney commented:
ok, I will take this up as soon as I get out of DLL hell with the "nupic.cpp as shared library" PR

dkeeney commented:
Ok, I am back working on SPRegion.   I have a question.  

The way it was working was:
- The inputDimensions argument to SP constructor is set by whatever the incoming buffer size is (flattened to one dimension). This is set by the upstream output (which may have allowed dimensions).

- The columnDimensions  argument to SP constructor was set using the width of the output Buffer which is set by the Link by calling getNodeOutputElementCount("bottomUpOut") which returns the 'columnCnt' parameter (one dimension).  No dimensions are allowed to be set on the SP.  


But now that we have the SDR for both input and output, what I would like to do is enable dimensions for SP within the NetworkAPI.
what I will do is:
- pass in the full dimensions from the incoming SDR as the inputDimensions constructor argument.
- Allow dimensions to be set on the SP but make them optional.
- in Link, If dimensions are not set, use the columnCnt parameter from configuration (as one dimension).
- in Link, use the resulting dimensions to create the output SDR (bottomUpOut") 
- in SPRegion, use the dimensions as the columnDimensions argument passed in to SP.

This will not change the API as long as we allow dimensions to be optional.  Any existing code will continue to work.  I will proceed with this assumption unless I hear objections.

dkeeney commented:
@ctrl-z-9000-times  The unit_tests are now failing in appveyor.  It might be due to some merge problem that messed up something.  Can you check this?  
```
[ RUN      ] SdrTest.TestCallbacks
ERR:  CHECK FAILED: "index < callbacks.size()" SparseDistributedRepresentation::removeCallback, Invalid Handle! [c:\projects\nupic-cpp\src\nupic\types\sdr.cpp line 428]
ERR:  CHECK FAILED: "index < callbacks.size()" SparseDistributedRepresentation::removeCallback, Invalid Handle! [c:\projects\nupic-cpp\src\nupic\types\sdr.cpp line 428]
ERR:  CHECK FAILED: "index < callbacks.size()" SparseDistributedRepresentation::removeCallback, Invalid Handle! [c:\projects\nupic-cpp\src\nupic\types\sdr.cpp line 428]
```

ctrl-z-9000-times commented:
> Can you check this?

It passes now...  I have no idea why this would suddenly happen and then disappear.  Nothing in that unit-test or area of code has changed recently.

dkeeney commented:
> I have no idea why this would suddenly happen and then disappear.

very strange.  That kind of thing happens if something did not get initialized but I did not see anything like that in the code.  

Anyway, review my PR when you get a chance.

ctrl-z-9000-times commented on src/nupic/regions/ScalarSensor.hpp:
```diff
@@ -70,7 +70,18 @@ class ScalarSensor : public RegionImpl {
+    UInt32 w;
+    bool periodic;
+    bool clipInput;
+  } params_;
```
Very nice!  This is good organization, and C++ even saves/loads the struct for you.

breznak replied:
for a discussion in a new issue, with these params, I had this idea I'd like to implement eventually: 
- each class implements (has a ) struct params_ member instead of its private/public fiends. 
- makes serialization easier
- has constructor that takes JSON
- we could remove (from API) all the public getters/setters for individual values, instead use
  - `get("columnSize"), set("isGlobal", false)`

breznak commented on src/nupic/engine/Input.hpp:
```diff
@@ -240,7 +240,7 @@ class Input {
 
   bool hasIncomingLinks() { return !links_.empty(); }
-
+  bool hasSplitterMap() { return !splitterMap_.empty(); }
```
is this part of a (public) API now?

dkeeney replied:
Don't worry, this gets removed in the next PR.  I added this so that if someone was using a splitterMap the Link class would make a copy of the SDR rather than share it when propagating buffers..

breznak commented on src/nupic/engine/Input.hpp:
```diff
@@ -260,6 +260,26 @@ class Input {
    * mutable because getSplitterMap() is const and logically
    * getting the splitter map doesn't change the Input
+   *
+   * What is a SplitterMap?
```
Thanks! These detailed explanations are very valuable for future-proof and new developers! :+1:

breznak commented on src/nupic/engine/Region.cpp:
```diff
@@ -61,13 +58,15 @@ Region::Region(std::string name, const std::string &nodeType,
 
+  // Dimensions start off as don't care but can be changed by user. dek 02/2019
+  // singleNodeOnly is ignored.
+  //dims_.push_back(1);
```
is this comment valid?

dkeeney replied:
Yes, I am trying to change the behavior a little.  Let me try to explain this...

The original behavior is that if the following line is in the Spec (and all region implementations do except TestNode)
```
ns->singleNodeOnly = true;
```
then the dimension is set to `dim = {1};` which means the output buffer is single dimension and the Link must ask the region implementation how wide it is (by calling getNodeOutputElementCount( )). So basically the region's configuration parameters determine the width of the output buffer and the user cannot call region->setDimension(dim) to set a dimension.

What I wanted to do was allow a user to set dimensions on a region anyway.  That value would be placed in the SDR when the Link creates the output SDR.  The dimensions on an input SDR would be determined by the output to which it is connected.

This is not perfect, the region->setDimension(dim) will affect all outputs of a region and I think the user should be able specify which output on a region they want the dimension to be set on.

I have been struggling with this.  I have been thinking that perhaps we should have a discussion about what a dimension means and do we need it at all and if so, do we need it to be set outside of an algorithm.  See Issue #273.

breznak commented on src/nupic/engine/Region.cpp:
```diff
@@ -61,13 +58,15 @@ Region::Region(std::string name, const std::string &nodeType,
   // can immediately set the dimensions.
   if (spec_->singleNodeOnly)
     dims_.push_back(1);
-  // else dims_ = []
```
if yes, this IF could be removed/commented out?

dkeeney replied:
You are correct. If I add this feature then the singleNodeOnly flag has no meaning.  The default would be we would always set the dimension to 1 which means single dimension with width determined by region parameters.

dkeeney replied:
I will fix that.

breznak commented on src/nupic/engine/Region.cpp:
```diff
@@ -303,11 +303,12 @@ void Region::setDimensions(Dimensions &newDims) {
               << newDims.toString() << " but region already has dimensions "
               << dims_.toString();
+    //NTA_THROW << "Attempt to set dimensions of region " << getName() << " to "
+    //          << newDims.toString() << " but region already initialized.";
```
rm please

dkeeney replied:
> rm please

Hmmm, it looks like I have that feature half on and half off.  Sorry.  I will fix this

breznak commented on src/nupic/ntypes/ArrayBase.cpp:
```diff
@@ -277,27 +277,27 @@ bool ArrayBase::has_buffer() const { return (buffer_.get() != nullptr); }
 void ArrayBase::convertInto(ArrayBase &a, size_t offset, size_t maxsize) const {
   if (maxsize == 0)
-    maxsize = getMaxElementsCount() + offset;
+    maxsize = getCount() + offset;
```
can this not change? The original capacity/MaxElementCount() seemed more fit?

dkeeney replied:
The problem I was trying to solve by changing this was that in a Fan-In condition I need to set the buffer size to the combined size of all of the buffers that will contribute.  That is the maxsize that is being passed in.  But if it is not a Fan-In condition (maxsize == 0) and I need to allocate a buffer then I should only allocate what I need, not the capacity of the source buffer.

breznak commented on src/nupic/regions/SPRegion.cpp:
```diff
@@ -0,0 +1,945 @@
+
+#define VERSION 1 // version for streaming serialization format
+
+namespace nupic {
```
are all of these directly in ::nupic? 
Not in this PR, but would be nice to introduce `::nupic::networkapi` namespace for these

dkeeney replied:
yes, everything is under nupic namespace.

breznak commented on src/nupic/regions/SPRegion.cpp:
```diff
@@ -0,0 +1,945 @@
+  // If there are more than one input link (FAN-IN), the input buffer will be the
+  // concatination of all incomming buffers.  
+  Input *in = getInput("bottomUpIn");
+  if (!in->hasIncomingLinks())
```
can in be null?

dkeeney replied:
Only if the Spec is wrong, but yes we should check.  good catch.
I added a check.

breznak commented on src/nupic/regions/SPRegion.cpp:
```diff
@@ -0,0 +1,945 @@
+  NTA_ASSERT(sp_) << "SP not initialized";
+
+  // BOTTOM-UP compute mode
+  args_.iter++;
```
is this part of the SPRegion, doesn't SP have its own iter, which we can use?

dkeeney replied:
Hmmm, interesting....this is not used for anything :)
I think in the original .py code it was used for something that did not apply to SP.

I will remove it.

breznak commented on src/nupic/regions/SPRegion.cpp:
```diff
@@ -0,0 +1,945 @@
+
+
+
+Spec *SPRegion::createSpec() {
```
maybe move this method to hpp? 
It's lot of lines, mostly doc, so it would fit there. (for your consideration only)

breznak commented on src/nupic/regions/SPRegion.cpp:
```diff
@@ -0,0 +1,945 @@
+////////////////////////////////////////////////////////////////////////
+
+UInt32 SPRegion::getParameterUInt32(const std::string &name, Int64 index) {
+  switch (name[0]) {
```
assert name.size()

dkeeney replied:
> assert name.size()

got it

breznak commented on src/nupic/regions/SPRegion.hpp:
```diff
@@ -0,0 +1,130 @@
+#include <nupic/algorithms/SpatialPooler.hpp>
+//----------------------------------------------------------------------
+
+using namespace nupic::algorithms::spatial_pooler;
```
move under the namespace nupic{ below, please

breznak replied:
or just using .. SpatialPooler, if possible (not namespace)

dkeeney replied:
ok.

breznak commented on src/nupic/regions/SPRegion.hpp:
```diff
@@ -0,0 +1,130 @@
+
+    // Used by RegionImplFactory to create and cache
+    // a nodespec. Ownership is transferred to the caller.
+    static Spec* createSpec();
```
does this have to be static? 
Can you make it a smart ptr?

dkeeney replied:
yes, it must be static.  It must be read before the region implementation is created.

breznak commented on src/nupic/regions/TestNode.cpp:
```diff
@@ -143,6 +143,8 @@ Spec *TestNode::createSpec() {
   ns->description = "TestNode. Used as a plain simple plugin Region for unit tests only. "
       "This is not useful for any real applicaton.";
 
+  ns->singleNodeOnly = false; // this means we allow dimensions;
```
drop singleNodeOnly altogether?

dkeeney replied:
> drop singleNodeOnly altogether?

Yes,  But hanging on to it until we are sure this is what we want.  See discussion on dimensions.

breznak commented on src/nupic/types/Sdr.cpp:
```diff
@@ -407,6 +407,7 @@ namespace nupic {
         // Consume the end marker.
         inStream >> marker;
         NTA_CHECK( marker == "~SDR" );
+        inStream.ignore(1);  // skip past endl.
```
maybe not, so we'd detect broken stream, saves?

dkeeney replied:
I had to add this because without it my load( ) was broken. :)

ctrl-z-9000-times replied:
+1 This is good.  Method `SDR.save()` always appends a newline to the stream which this correctly chomps off.

dkeeney commented:
I think I addressed all of the concerns.  I backed out the change in behavior of Dimensions because in the light of the discussion about Dimensions #273 I think I need to do that in a separate PR.

breznak commented:
Thanks for making it with this PR, @dkeeney ! Having SP,TM regions in networkAPI would be a great addition to this repo :1st_place_medal:
